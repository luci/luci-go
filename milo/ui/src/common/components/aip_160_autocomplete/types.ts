// Copyright 2025 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { UseQueryOptions } from '@tanstack/react-query';
import { ReactNode } from 'react';

import { NonNullableProps } from '@/generic_libs/types';

/**
 * The definition of a single field (of a larger object).
 */
export interface FieldDef {
  readonly staticFields?: StaticFieldsSchema;
  readonly dynamicFields?: DynamicFieldsSchema;

  /**
   * A function that returns the values to be suggested to the users.
   */
  readonly getValues?: (partial: string) => readonly ValueDef[];
  /**
   * A function that returns the react-query to be used to retrieve value
   * suggestions.
   */
  // We don't care about the shape of the data returned from the `queryFn`.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  readonly fetchValues?: FetchValuesFn<any>;
}

/**
 * Defines the static fields (of a larger object).
 */
export interface StaticFieldsSchema {
  /**
   * A map of fields.
   *
   * Fields that includes the user input as a substring (case-insensitive) will
   * be suggested to the users.
   * Exact match (case-sensitive) are not suggested.
   */
  readonly [field: string]: FieldDef;
}

/**
 * Defines the dynamic fields (of a larger object).
 */
export interface DynamicFieldsSchema {
  readonly getKeys?: (partial: string) => readonly ValueDef[];
  readonly getValues?: (key: string, partial: string) => readonly ValueDef[];
}

/**
 * The definition of a single value (to be used in a field).
 */
export interface ValueDef {
  readonly text: string;
  readonly unselectable?: boolean;
  /**
   * HTML (React) representation of the value.
   *
   * Defaults to `text`.
   */
  readonly display?: ReactNode;
  /**
   * Explanation attached to this value.
   */
  readonly explanation?: ReactNode;
}

/**
 * A function that returns the react-query option to be used to retrieve value
 * suggestions.
 */
export type FetchValuesFn<T> = (partial: string) => ValueQueryOptions<T>;

/**
 * A react-query option to be used to retrieve value suggestions.
 */
export type ValueQueryOptions<T> =
  | Omit<UseQueryOptions<readonly ValueDef[]>, 'select'>
  // We don't care about the shape of the data returned from the `queryFn` when
  // `select` is specified.
  | NonNullableProps<UseQueryOptions<T, Error, readonly ValueDef[]>, 'select'>;

/**
 * An AIP-160 filter autocomplete suggestion
 */
export interface Suggestion {
  /**
   * Text representation of the suggestion.
   */
  readonly text: string;
  /**
   * HTML (React) representation of the suggestion.
   *
   * Defaults to `text`.
   */
  readonly display?: ReactNode;
  /**
   * Explanation attached to this suggestion.
   */
  readonly explanation?: ReactNode;
  /**
   * Apply the suggestion to the input.
   *
   * When called, returns the updated input and cursor position if the
   * suggestion is applied.
   */
  readonly apply: () => readonly [string, number];
}
