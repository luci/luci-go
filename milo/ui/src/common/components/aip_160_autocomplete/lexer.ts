// Copyright 2025 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Enum for token kinds.
 */
export enum TokenKind {
  Whitespace = 'WHITESPACE',
  Paren = 'PAREN',
  Comma = 'COMMA',
  Comparator = 'COMPARATOR',
  Keyword = 'KEYWORD',
  /**
   * `-` with an expression following it.
   */
  Negate = 'NEGATE',
  QualifiedFunction = 'QUALIFIED_FUNCTION',
  QualifiedFieldOrValue = 'QUALIFIED_FIELD_OR_VALUE',
  /**
   * Quoted single line string.
   */
  String = 'STRING',
  /**
   * Quoted single line string but without the closing quote.
   * Terminates before the first line break or at the end of the input.
   */
  UnclosedString = 'UNCLOSED_STRING',
  /**
   * Invalid char. Allow us to process the rest of the tokens even when we
   * encounter an unexpected one.
   */
  InvalidChar = 'INVALID_CHAR',
}

const TOKEN_KINDS = Object.values(TokenKind);

/**
 * LEXER_RE has one capture group for each kind of token that can be lexed.
 * The ith capturing group represents a token of the ith kind in `TokenKind`.
 */
const LEXER_RE = new RegExp(
  /^/.source +
    [
      // Whitespace
      /\s+/,
      // Paren
      /[()]/,
      // Comma
      /,/,
      // Comparator
      /<|<=|=|>=|>|!=|:/,
      // Keyword
      /AND\b|OR\b|NOT\b/,
      // Negate
      /-(?!\s)/,
      // QualifiedFunction
      /[\w._-]+\(/,
      // QualifiedField
      /[\w._-]+/,
      // String
      /'(?:[^"\\\n]|\\.)*'|"(?:[^"\\\n]|\\.)*"/,
      // UnclosedString
      /'(?:[^'\\\n]|\\.)*(?=\n|$)|"(?:[^"\\\n]|\\.)*(?=\n|$)/,
      // InvalidChar
      /.|\n/,
    ]
      .map((r) => `(${r.source})`)
      .join('|'),
);

export interface Token {
  readonly index: number;
  readonly startPos: number;
  readonly kind: TokenKind;
  readonly text: string;
}

export class Lexer {
  private cachedTokens: Token[] = [];
  private unprocessedInput: string;
  private processedLen = 0;

  constructor(private input: string) {
    this.unprocessedInput = this.input;
  }

  getAllTokens(): readonly Token[] {
    while (this.unprocessedInput !== '') {
      this.processNextToken();
    }
    return this.cachedTokens;
  }

  /**
   * Get token(s) at the specified cursor position. Return
   *  * two tokens: if the cursor divide two tokens (e.g. `text|<=`).
   *  * one token: if the cursor is in the middle of a token (e.g. `te|xt<=`)
   *  * zero token: if the cursor is not placed on any token (e.g. `text<=  |`)
   */
  getAtPosition(
    cursorPos: number,
  ): readonly [Token, Token] | readonly [Token] | readonly [] {
    while (this.unprocessedInput !== '' && this.processedLen <= cursorPos + 1) {
      this.processNextToken();
    }

    // The text is empty.
    if (this.processedLen === 0) {
      return [];
    }

    // The cursor position is not in the range of the text [0, text_length].
    if (cursorPos < 0 || cursorPos > this.processedLen) {
      return [];
    }

    // Find the LAST token that starts before the cursor.
    let tokenIndex = -1;
    for (let i = 0; i < this.cachedTokens.length; ++i) {
      if (this.cachedTokens[i].startPos >= cursorPos) {
        break;
      }
      tokenIndex = i;
    }

    // `prevToken !== null` is always true because the previous two early
    // returns guarantee there's a token that starts before the cursor.
    const prevToken = this.cachedTokens[tokenIndex];
    const nextToken = this.cachedTokens[tokenIndex + 1] as Token | undefined;

    // If the next token starts at the cursor position, the cursor must divide
    // two tokens (e.g. `${prevToken}|${nextToken}`). Return both of them.
    if (nextToken?.startPos === cursorPos) {
      return [prevToken, nextToken];
    }

    return [prevToken];
  }

  /**
   * Get the last token with an index smaller than the specified index.
   * Useful for retrieving the previous token of a given token (index).
   *
   * @param allowWS When set the true, include `Whitespace` token. Defaults to
   *   false.
   */
  getBeforeIndex(beforeIndex: number, allowWS = false): Token | null {
    while (
      this.unprocessedInput !== '' &&
      this.cachedTokens.length < beforeIndex
    ) {
      this.processNextToken();
    }

    const searchStart = Math.min(beforeIndex - 1, this.cachedTokens.length - 1);
    for (let i = searchStart; i >= 0; --i) {
      const token = this.cachedTokens[i];
      if (!allowWS && token.kind === TokenKind.Whitespace) {
        continue;
      }
      return token;
    }
    return null;
  }

  /**
   * Get the first token with an index greater than the specified index.
   * Useful for retrieving the next token of a given token (index).
   *
   * @param allowWS When set the true, include `Whitespace` token. Defaults to
   *   false.
   */
  getAfterIndex(afterIndex: number, allowWS = false): Token | null {
    const searchStart = Math.max(afterIndex + 1, 0);
    let i = searchStart;
    for (;;) {
      while (i >= this.cachedTokens.length && this.unprocessedInput !== '') {
        this.processNextToken();
      }
      if (i >= this.cachedTokens.length) {
        return null;
      }

      const token = this.cachedTokens[i];
      ++i;
      if (!allowWS && token.kind === TokenKind.Whitespace) {
        continue;
      }
      return token;
    }
  }

  /**
   * Process the next token and append it to `this.cachedTokens`.
   */
  private processNextToken() {
    if (this.unprocessedInput === '') {
      return;
    }

    // Match the next token string.
    const matches = LEXER_RE.exec(this.unprocessedInput);
    if (!matches) {
      throw new Error(
        `invariant violated: unable to lex token from '${this.input}' at position ${this.processedLen}`,
      );
    }
    const startPos = this.processedLen;
    this.processedLen += matches[0].length;
    this.unprocessedInput = this.unprocessedInput.slice(matches[0].length);

    // Map the token string to a token kind.
    const kindIter = TOKEN_KINDS.entries();
    for (const [i, kind] of kindIter) {
      if (matches[i + 1] === undefined) {
        continue;
      }
      this.cachedTokens.push({
        index: this.cachedTokens.length,
        startPos,
        kind,
        text: matches[i + 1],
      });
      return;
    }

    // Should never happen because we have a catch all token type.
    throw new Error(
      `invariant violated: unhandled lexer regexp match '${matches[TOKEN_KINDS.length]}'`,
    );
  }
}
