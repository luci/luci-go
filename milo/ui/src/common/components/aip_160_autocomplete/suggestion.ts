// Copyright 2025 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { useQuery } from '@tanstack/react-query';
import { useMemo } from 'react';

import { OptionDef } from '@/generic_libs/components/text_autocomplete';

import { Lexer, Token, TokenKind } from './lexer';
import { FieldDef, FieldsSchema, Suggestion } from './types';

export function useSuggestions(
  schema: FieldsSchema,
  input: string,
  cursorPos: number,
): readonly OptionDef<Suggestion>[] {
  const lexer = useMemo(() => new Lexer(input), [input]);
  const ctx = getSuggestionCtx(lexer, cursorPos);

  const optionsSync = useSuggestionsSync(schema, input, ctx);
  const optionsAsync = useSuggestionsAsync(schema, input, ctx);

  return [...optionsSync, ...optionsAsync];
}

/**
 * Token selection priority. Lower index -> higher priority.
 *
 * When the cursor is between two tokens, the one with higher priority will be
 * selected.
 */
const TOKEN_SELECTION_PRIORITY = Object.freeze([
  TokenKind.QualifiedFieldOrValue,
  TokenKind.String,
  TokenKind.UnclosedString,
  TokenKind.QualifiedFunction,
  TokenKind.Comparator,
  TokenKind.Paren,
  TokenKind.Comma,
  TokenKind.Keyword,
  TokenKind.Negate,
  TokenKind.InvalidChar,
  TokenKind.Whitespace,
  undefined,
]);

export type SuggestionContext =
  | SuggestionContextField
  | SuggestionContextValue
  | null;

export interface SuggestionContextField {
  readonly type: 'field';
  readonly fieldToken: Token;
}

export interface SuggestionContextValue {
  readonly type: 'value';
  readonly fieldToken: Token;
  readonly valueToken: Token;
}

export function getSuggestionCtx(
  lexer: Lexer,
  cursorPos: number,
): SuggestionContext {
  const tokens = lexer.getAtPosition(cursorPos);

  // Get the token at the cursor position.
  // When the cursor is between two tokens, select the one with higher priority.
  const selectedToken = [...tokens, undefined].sort(
    (a, b) =>
      TOKEN_SELECTION_PRIORITY.indexOf(a?.kind) -
      TOKEN_SELECTION_PRIORITY.indexOf(b?.kind),
  )[0];

  if (!selectedToken) {
    return null;
  }

  const prevToken1 = lexer.getBeforeIndex(selectedToken.index);
  const prevToken2 = prevToken1 ? lexer.getBeforeIndex(prevToken1.index) : null;

  // Suggest value if the current token can be interpreted as a value token.

  // Example: `field > val|ue`
  if (
    prevToken2?.kind === TokenKind.QualifiedFieldOrValue &&
    prevToken1?.kind === TokenKind.Comparator &&
    selectedToken.kind === TokenKind.QualifiedFieldOrValue
  ) {
    return {
      type: 'value',
      fieldToken: prevToken2,
      valueToken: selectedToken,
    };
  }

  // Example: `field > "val|ue"`
  if (
    prevToken2?.kind === TokenKind.QualifiedFieldOrValue &&
    prevToken1?.kind === TokenKind.Comparator &&
    selectedToken.kind === TokenKind.String
  ) {
    return {
      type: 'value',
      fieldToken: prevToken2,
      valueToken: selectedToken,
    };
  }

  // Example: `field > "val|ue`
  if (
    prevToken2?.kind === TokenKind.QualifiedFieldOrValue &&
    prevToken1?.kind === TokenKind.Comparator &&
    selectedToken.kind === TokenKind.UnclosedString
  ) {
    return {
      type: 'value',
      fieldToken: prevToken2,
      valueToken: selectedToken,
    };
  }

  // Example: `field >|`
  if (
    prevToken1?.kind === TokenKind.QualifiedFieldOrValue &&
    selectedToken.kind === TokenKind.Comparator
  ) {
    return {
      type: 'value',
      fieldToken: prevToken1,
      valueToken: {
        index: selectedToken.index + 1,
        startPos: selectedToken.startPos + selectedToken.text.length,
        kind: TokenKind.QualifiedFieldOrValue,
        text: '',
      },
    };
  }

  // Example: `field >  | `
  if (
    prevToken2?.kind === TokenKind.QualifiedFieldOrValue &&
    prevToken1?.kind === TokenKind.Comparator &&
    selectedToken.kind === TokenKind.Whitespace
  ) {
    return {
      type: 'value',
      fieldToken: prevToken2,
      valueToken: {
        index: selectedToken.index,
        // Leave one space no matter how long the whitespace is.
        startPos: selectedToken.startPos + 1,
        kind: TokenKind.QualifiedFieldOrValue,
        text: '',
      },
    };
  }

  // Suggest field otherwise.

  // Example: fie|ld
  if (selectedToken.kind === TokenKind.QualifiedFieldOrValue) {
    return {
      type: 'field',
      fieldToken: selectedToken,
    };
  }

  return null;
}

function useSuggestionsSync(
  schema: FieldsSchema,
  input: string,
  ctx: SuggestionContext,
): readonly OptionDef<Suggestion>[] {
  const options: OptionDef<Suggestion>[] = [];
  if (ctx?.type === 'value') {
    options.push(
      ...suggestValue(schema, input, ctx.fieldToken, ctx.valueToken),
    );
  }
  if (ctx?.type === 'field') {
    options.push(...suggestField(schema, input, ctx.fieldToken));
  }
  return options;
}

function useSuggestionsAsync(
  schema: FieldsSchema,
  input: string,
  ctx: SuggestionContext,
): readonly OptionDef<Suggestion>[] {
  const valueQuery = (ctx &&
    ctx.type === 'value' &&
    getField(schema, ctx.fieldToken.text)?.fetchValues?.(
      ctx.valueToken.text,
    )) || {
    queryKey: ['always-disabled'],
    queryFn: () => [],
    enabled: false,
  };

  const { data, isError, error, isLoading, isFetching } = useQuery(valueQuery);
  if (isError) {
    throw error;
  }

  if (isLoading && !isFetching) {
    return [];
  }
  // This is already covered by the case above. Add it for type narrowing.
  if (ctx?.type !== 'value') {
    return [];
  }

  if (isLoading) {
    return [
      {
        id: 'loading',
        value: {
          text: 'Loading...',
          // Doesn't matter. This option is not selectable.
          apply: () => ['', 0] as const,
        },
        unselectable: true,
      },
    ];
  }

  return (
    data?.map((valueDef) => ({
      id: valueDef.text,
      value: {
        text: valueDef.text,
        apply: () => {
          const prefixEnd = ctx.valueToken.startPos;
          const suffixStart =
            ctx.valueToken.startPos + ctx.valueToken.text.length;
          return [
            input.slice(0, prefixEnd) +
              valueDef.text +
              input.slice(suffixStart),
            prefixEnd + valueDef.text.length,
          ] as const;
        },
      },
    })) || []
  );
}

function suggestField(
  schema: FieldsSchema,
  input: string,
  fieldToken: Token,
): readonly OptionDef<Suggestion>[] {
  const lastDotIndex = fieldToken.text.lastIndexOf('.');
  const prev = fieldToken.text.slice(0, Math.max(lastDotIndex, 0));
  const suffix = fieldToken.text.slice(lastDotIndex + 1);
  const suffixLower = suffix.toLowerCase();
  const targetSchema =
    prev === '' ? schema : getField(schema, prev)?.fields || {};

  return Object.keys(targetSchema)
    .filter(
      (name) => name.toLowerCase().includes(suffixLower) && name !== suffix,
    )
    .map((name) => ({
      id: name,
      value: {
        text: name,
        apply: () => {
          const prefixEnd = fieldToken.startPos + lastDotIndex + 1;
          const suffixStart =
            prefixEnd + fieldToken.text.length - lastDotIndex - 1;
          return [
            input.slice(0, prefixEnd) + name + input.slice(suffixStart),
            prefixEnd + name.length,
          ] as const;
        },
      },
    }));
}

function suggestValue(
  schema: FieldsSchema,
  input: string,
  fieldToken: Token,
  valueToken: Token,
): readonly OptionDef<Suggestion>[] {
  const field = getField(schema, fieldToken.text);
  return (
    field?.getValues?.(valueToken.text).map((valueDef) => ({
      id: valueDef.text,
      value: {
        text: valueDef.text,
        apply: () => {
          const prefixEnd = valueToken.startPos;
          const suffixStart = valueToken.startPos + valueToken.text.length;
          return [
            input.slice(0, prefixEnd) +
              valueDef.text +
              input.slice(suffixStart),
            prefixEnd + valueDef.text.length,
          ] as const;
        },
      },
    })) || []
  );
}

function getField(schema: FieldsSchema, fieldPath: string): FieldDef | null {
  const fieldNames = fieldPath.split('.');

  let pivot = schema;
  for (let i = 0; i < fieldNames.length - 1; ++i) {
    const fieldName = fieldNames[i];
    const fieldDef = schema[fieldName];
    if (!fieldDef?.fields) {
      return null;
    }
    pivot = fieldDef.fields;
  }
  return pivot[fieldNames.at(-1)!] ?? null;
}
