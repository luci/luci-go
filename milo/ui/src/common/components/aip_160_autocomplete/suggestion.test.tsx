// Copyright 2025 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { cleanup, renderHook } from '@testing-library/react';
import { act } from 'react';

import { deferred } from '@/generic_libs/tools/utils';

import { Lexer } from './lexer';
import { useSuggestions } from './suggestion';
import { getSuggestionCtx } from './suggestion';
import { FieldDef } from './types';

describe('useSuggestions', () => {
  const queryClient = new QueryClient();
  const Wrapper = ({ children }: { children: React.ReactNode }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  );

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
    cleanup();
  });

  it('should suggest fields', () => {
    const schema: FieldDef = {
      fields: {
        project: {},
        projectWithSuffix: {},
        builder: {},
      },
    };

    const { result } = renderHook(() => useSuggestions(schema, 'projec', 3), {
      wrapper: Wrapper,
    });
    expect(result.current).toEqual([
      {
        id: 'project',
        value: {
          text: 'project',
          display: expect.anything(),
          apply: expect.any(Function),
        },
      },
      {
        id: 'projectWithSuffix',
        value: {
          text: 'projectWithSuffix',
          display: expect.anything(),
          apply: expect.any(Function),
        },
      },
    ]);
    expect(result.current[0]!.value.apply()).toEqual(['project', 7]);
    expect(result.current[1]!.value.apply()).toEqual(['projectWithSuffix', 17]);
  });

  it('should suggest nested fields', () => {
    const schema: FieldDef = {
      fields: {
        build: {
          fields: {
            builderId: {
              fields: {
                project: {},
              },
            },
            // Should not be suggested when `builderId.` prefix exists.
            projectWithSuffix: {},
          },
        },
      },
    };

    const { result } = renderHook(
      () => useSuggestions(schema, 'build.builderId.pro', 3),
      {
        wrapper: Wrapper,
      },
    );
    expect(result.current).toEqual([
      {
        id: 'project',
        value: {
          text: 'project',
          display: expect.anything(),
          apply: expect.any(Function),
        },
      },
    ]);
    expect(result.current[0]!.value.apply()).toEqual([
      'build.builderId.project',
      23,
    ]);
  });

  it('should not suggest anything when the parent field does not exist', () => {
    const schema: FieldDef = {
      fields: {
        builderId: {
          fields: {
            project: {},
          },
        },
        projectWithSuffix: {},
      },
    };

    const { result } = renderHook(
      () => useSuggestions(schema, 'builder.pro', 3),
      {
        wrapper: Wrapper,
      },
    );
    expect(result.current).toEqual([]);
  });

  it('should be case insensitive when suggesting fields', () => {
    const schema: FieldDef = {
      fields: {
        Project: {},
        project: {},
      },
    };

    const { result } = renderHook(() => useSuggestions(schema, 'project', 3), {
      wrapper: Wrapper,
    });
    expect(result.current).toEqual([
      {
        id: 'Project',
        value: {
          text: 'Project',
          display: expect.anything(),
          apply: expect.any(Function),
        },
      },
    ]);
  });

  it('should not suggest field with the exact match', () => {
    const schema: FieldDef = {
      fields: {
        project: {},
        projectWithSuffix: {},
      },
    };

    const { result } = renderHook(() => useSuggestions(schema, 'project', 3), {
      wrapper: Wrapper,
    });
    expect(result.current).toEqual([
      {
        id: 'projectWithSuffix',
        value: {
          text: 'projectWithSuffix',
          display: expect.anything(),
          apply: expect.any(Function),
        },
      },
    ]);
  });

  it('should suggest values', () => {
    const schema: FieldDef = {
      fields: {
        project: {
          getValues: (partial) =>
            ['chromium', 'chromeos']
              .filter((text) => text.includes(partial))
              .map((text) => ({ text })),
        },
      },
    };

    const { result } = renderHook(
      () => useSuggestions(schema, 'project = chrom', 13),
      { wrapper: Wrapper },
    );
    expect(result.current).toEqual([
      {
        id: 'chromium',
        value: {
          text: 'chromium',
          apply: expect.any(Function),
        },
      },
      {
        id: 'chromeos',
        value: {
          text: 'chromeos',
          apply: expect.any(Function),
        },
      },
    ]);
    expect(result.current[0]!.value.apply()).toEqual([
      'project = chromium',
      18,
    ]);
    expect(result.current[1]!.value.apply()).toEqual([
      'project = chromeos',
      18,
    ]);
  });

  it('should suggest value of a nested field', () => {
    const schema: FieldDef = {
      fields: {
        build: {
          fields: {
            builderId: {
              fields: {
                project: {
                  getValues: (partial) =>
                    ['chromium', 'chromeos']
                      .filter((text) => text.includes(partial))
                      .map((text) => ({ text })),
                },
              },
            },
          },
        },
      },
    };

    const { result } = renderHook(
      () => useSuggestions(schema, 'build.builderId.project = chrom', 28),
      { wrapper: Wrapper },
    );
    expect(result.current).toEqual([
      {
        id: 'chromium',
        value: {
          text: 'chromium',
          apply: expect.any(Function),
        },
      },
      {
        id: 'chromeos',
        value: {
          text: 'chromeos',
          apply: expect.any(Function),
        },
      },
    ]);
    expect(result.current[0]!.value.apply()).toEqual([
      'build.builderId.project = chromium',
      34,
    ]);
    expect(result.current[1]!.value.apply()).toEqual([
      'build.builderId.project = chromeos',
      34,
    ]);
  });

  it('should suggest values asynchronously', async () => {
    const [blocker, resolveBlocker] = deferred();
    const schema: FieldDef = {
      fields: {
        project: {
          fetchValues: (partial: string) => ({
            queryKey: [partial],
            queryFn: async () => {
              await blocker;
              return ['chromium', 'chromeos']
                .filter((text) => text.includes(partial) && text !== partial)
                .map((text) => ({ text }));
            },
          }),
        },
      },
    };

    const { result } = renderHook(
      () => useSuggestions(schema, 'project = chromi', 12),
      { wrapper: Wrapper },
    );

    // Initially, should show loading.
    expect(result.current).toEqual([
      {
        id: 'loading',
        value: {
          text: 'Loading...',
          display: expect.anything(),
          apply: expect.any(Function),
        },
        unselectable: true,
      },
    ]);

    resolveBlocker();
    await act(() => jest.runOnlyPendingTimersAsync());

    expect(result.current).toEqual([
      {
        id: 'chromium',
        value: {
          text: 'chromium',
          apply: expect.any(Function),
        },
      },
    ]);
    expect(result.current[0]!.value.apply()).toEqual([
      'project = chromium',
      18,
    ]);
  });
});

describe('getSuggestionCtx', () => {
  it.each([
    // Cursor on field.
    ['|project'],
    ['pro|ject'],
    ['project|'],

    // Cursor on value.
    ['project = |chromium'],
    ['project = chro|mium'],
    ['project = chromium|'],

    // Cursor on value no whitespace.
    ['project=|chromium'],
    ['project=chro|mium'],
    ['project=chromium|'],

    // Cursor after value.
    ['project=chromium |'],

    // Cursor after comparator.
    ['project = | '],

    // Cursor after comparator with longer whitespace.
    ['project =   | '],

    // Cursor on field with following value.
    ['|project=chromium'],
    ['pro|ject=chromium'],
    ['project|=chromium'],

    // Cursor on whitespace after field.
    ['project |'],
    ['project | another_project'],

    // Quoted string.
    ['project = "chromi|um"'],
    // Unclosed quoted string.
    ['project = "chromi|um'],
  ])('inputWithCursor: %s', (inputWithCursor) => {
    const input = inputWithCursor.replace('|', '');
    const cursorPos = inputWithCursor.indexOf('|');
    const lexer = new Lexer(input);
    expect(getSuggestionCtx(lexer, cursorPos)).toMatchSnapshot();
  });
});
