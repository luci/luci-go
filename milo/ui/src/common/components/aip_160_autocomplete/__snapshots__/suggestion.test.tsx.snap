// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSuggestionCtx inputWithCursor: |project 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "field",
}
`;

exports[`getSuggestionCtx inputWithCursor: |project=chromium 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "field",
}
`;

exports[`getSuggestionCtx inputWithCursor: pro|ject 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "field",
}
`;

exports[`getSuggestionCtx inputWithCursor: pro|ject=chromium 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "field",
}
`;

exports[`getSuggestionCtx inputWithCursor: project =   |  1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "value",
  "valueToken": {
    "index": 3,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 10,
    "text": "",
  },
}
`;

exports[`getSuggestionCtx inputWithCursor: project = "chromi|um 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "value",
  "valueToken": {
    "index": 4,
    "kind": "UNCLOSED_STRING",
    "startPos": 10,
    "text": ""chromium",
  },
}
`;

exports[`getSuggestionCtx inputWithCursor: project = "chromi|um" 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "value",
  "valueToken": {
    "index": 4,
    "kind": "STRING",
    "startPos": 10,
    "text": ""chromium"",
  },
}
`;

exports[`getSuggestionCtx inputWithCursor: project = |  1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "value",
  "valueToken": {
    "index": 3,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 10,
    "text": "",
  },
}
`;

exports[`getSuggestionCtx inputWithCursor: project = |chromium 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "value",
  "valueToken": {
    "index": 4,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 10,
    "text": "chromium",
  },
}
`;

exports[`getSuggestionCtx inputWithCursor: project = chro|mium 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "value",
  "valueToken": {
    "index": 4,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 10,
    "text": "chromium",
  },
}
`;

exports[`getSuggestionCtx inputWithCursor: project = chromium| 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "value",
  "valueToken": {
    "index": 4,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 10,
    "text": "chromium",
  },
}
`;

exports[`getSuggestionCtx inputWithCursor: project | 1`] = `null`;

exports[`getSuggestionCtx inputWithCursor: project | another_project 1`] = `null`;

exports[`getSuggestionCtx inputWithCursor: project=|chromium 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "value",
  "valueToken": {
    "index": 2,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 8,
    "text": "chromium",
  },
}
`;

exports[`getSuggestionCtx inputWithCursor: project=chro|mium 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "value",
  "valueToken": {
    "index": 2,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 8,
    "text": "chromium",
  },
}
`;

exports[`getSuggestionCtx inputWithCursor: project=chromium | 1`] = `null`;

exports[`getSuggestionCtx inputWithCursor: project=chromium| 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "value",
  "valueToken": {
    "index": 2,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 8,
    "text": "chromium",
  },
}
`;

exports[`getSuggestionCtx inputWithCursor: project| 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "field",
}
`;

exports[`getSuggestionCtx inputWithCursor: project|=chromium 1`] = `
{
  "fieldToken": {
    "index": 0,
    "kind": "QUALIFIED_FIELD_OR_VALUE",
    "startPos": 0,
    "text": "project",
  },
  "type": "field",
}
`;
